src/Language/PureScript/Linter.hs

```diff
@@ -129,3 +130,158 @@ lint (Module _ _ mn ds _) = censor (addHint (ErrorInModule mn)) $ mapM_ lintDecl
+                allExprs = concatMap unguard gexprs
+            in
+                removeAndWarn ss bindNewNames $ combineAll $ map (go ss) allExprs
+      in
+      combineAll $ map (go ss) vs ++ map f alts
+
+    go ss (TypedValue _ v1 _) = go ss v1
+    go ss (Do _ es) = doElts ss es Nothing
+    go ss (Ado _ es v1) = doElts ss es (Just v1)
+
+    go ss (Literal _ (ArrayLiteral es)) = combineAll $ map (go ss) es
+    go ss (Literal _ (ObjectLiteral oo)) = combineAll $ map (go ss . snd) oo
+
+    go _ (PositionedValue ss' _ v1) = go ss' v1
+
+    go _ _ = (mempty, mempty)
+
+    doElts :: SourceSpan -> [DoNotationElement] -> Maybe Expr -> (S.Set Ident, MultipleErrors)
+    doElts ss' (DoNotationValue e : rest) v = go ss' e `combine` doElts ss' rest v
+    doElts ss' (DoNotationBind binder e : rest) v =
+      let bindNewNames = S.fromList (binderNames binder)
+      in go ss' e `combine` removeAndWarn ss' bindNewNames (doElts ss' rest v)
+
+    doElts ss' (DoNotationLet ds : rest) v =
+      let letNewNames = S.fromList $ concatMap declIdents ds
+          declRes = foldr1 combine (map underDecl ds)
+      in removeAndWarn ss' letNewNames $ declRes `combine` doElts ss' rest v
+    doElts _ (PositionedDoNotationElement ss'' _ e : rest) v = doElts ss'' (e : rest) v
+    doElts ss' [] (Just e) = go ss' e `combine` (rebindable, mempty)
+    doElts _ [] Nothing = (rebindable, mempty)
+
+    declIdents :: Declaration -> [Ident]
+    declIdents (ValueDecl _ ident _ _ _) = [ident]
+    declIdents (BoundValueDeclaration _ binders _) = binderNames binders
+    declIdents _ = []
+
+    -- let f x = e  -- check the x in e (but not the f)
+    underDecl d@(ValueDecl _ _ _ binders gexprs) =
+      let bindNewNames = S.fromList (concatMap binderNames binders)
+          allExprs = concatMap unguard gexprs
+          ss = declSourceSpan d
+      in
+          removeAndWarn ss bindNewNames $ foldr1 combine $ map (go ss) allExprs
+    -- let {x} = e  -- no binding to check inside e
+    underDecl d@(BoundValueDeclaration _ _ expr) =
+      go (declSourceSpan d) expr
+    underDecl _ = (mempty, mempty)
+
+    unguard (GuardedExpr guards expr) = map unguard' guards ++ [expr]
+    unguard' (ConditionGuard ee) = ee
+    unguard' (PatternGuard _ ee) = ee
+
+    removeAndWarn :: SourceSpan -> S.Set Ident -> (S.Set Ident, MultipleErrors) -> (S.Set Ident, MultipleErrors)
+    removeAndWarn ss newNames (used, errors) =
+      let filteredUsed = used `S.difference` newNames
+          warnUnused = S.filter (not . Text.isPrefixOf "_" . runIdent) (newNames `S.difference` used)
+          combinedErrors = if not $ S.null warnUnused then errors <> (mconcat $ map (errorMessage' ss . UnusedName) $ S.toList warnUnused) else errors
+      in
+        (filteredUsed, combinedErrors)
+
+    combine ::
+      (S.Set Ident, MultipleErrors) ->
+      (S.Set Ident, MultipleErrors) ->
+      (S.Set Ident, MultipleErrors)
+    combine (a, b) (c, d) = (S.union a c, b <> d)
```

Comment:
I think this is just `<>`? Or did you avoid that deliberately for readability?

Reply:
I imagine this is like this because of the other version in this file which is `(S.intersection a c, b <> d)` - I think this code originally worked the other way around so the instance wasn't right. I don't think infix `combine` is readable - I'll change it